src/app/clusters/microwave-oven-control-server/microwave-oven-control-server.cpp

```diff
@@ -0,0 +1,330 @@
+            reqCookTime = MicrowaveOvenControl::kDefaultCookTime;
+        }
+
+        if (PowerSetting.HasValue())
+        {
+            reqPowerSetting = PowerSetting.Value();
+        }
+        else
+        {
+            // set Microwave Oven cooking power to max power(default).
+            reqPowerSetting = mDelegate->GetMaxPower();
+        }
+        status = mDelegate->HandleSetCookingParametersCallback(reqCookMode, reqCookTime, reqPowerSetting);
+        goto exit;
+    }
+    else
+    {
+        status = Status::InvalidInState;
+        goto exit;
+    }
+
+exit:
+    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+}
+
+void Instance::HandleAddMoreTime(HandlerContext & ctx, const Commands::AddMoreTime::DecodableType & req)
+{
+    ChipLogDetail(Zcl, "MicrowaveOvenControl:  HandleAddMoreTime");
+    Status status;
+    uint8_t opState;
+
+    // Get Operational State instance
+    gOperationalStateInstance = GetOPInstance(mEndpointId);
+    if (gOperationalStateInstance == nullptr)
+    {
+        status = Status::InvalidInState;
+        ChipLogError(Zcl, "Microwave Oven Control: Operational State instance does not exist");
+        goto exit;
+    }
+
+    opState = gOperationalStateInstance->GetCurrentOperationalState();
+    if (opState == to_underlying(OperationalStateEnum::kStopped) || opState == to_underlying(OperationalStateEnum::kRunning) ||
+        opState == to_underlying(OperationalStateEnum::kPaused))
+    {
+        uint32_t addedCookTime;
+
+        addedCookTime = GetCookTime() + req.timeToAdd;
+        // if the added cooking time is greater than the max cooking time, the cooking time stay unchanged.
+        if (addedCookTime < kMaxCookTime)
+        {
+            status = mDelegate->HandleAddMoreTimeCallback(addedCookTime);
+            goto exit;
+        }
+        else
+        {
+            ChipLogError(Zcl, "Failed to set cookTime, cookTime value is out of range");
+            status = Status::ConstraintError;
+            goto exit;
+        }
+    }
+    else // operational state is in error
+    {
+        status = Status::InvalidInState;
+        goto exit;
+    }
+
+exit:
+    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+}
+
+bool IsCookTimeInRange(uint32_t cookTime)
+{
+    return (cookTime < MicrowaveOvenControl::kMinCookTime || cookTime > MicrowaveOvenControl::kMaxCookTime) ? false : true;
+}
+
+bool IsPowerSettingInRange(uint8_t powerSetting, uint8_t minCookPower, uint8_t maxCookPower)
+{
+    return (powerSetting < minCookPower || powerSetting > maxCookPower) ? false : true;
+}
+
+void SetOPInstance(EndpointId aEndpoint, OperationalState::Instance * aInstance)
+{
+    uint16_t ep = emberAfGetClusterServerEndpointIndex(aEndpoint, MicrowaveOvenControl::Id,
+                                                       EMBER_AF_MICROWAVE_OVEN_CONTROL_CLUSTER_SERVER_ENDPOINT_COUNT);
+    // if endpoint is found
+    if (ep < kMicrowaveOvenControlInstanceTableSize)
```

Comment:
This ties the endpoint IDs to the table size and assumes there are only microwave oven endpoints on the device. Please make the table contain entries of a struct of `EndpointId, Instance` so that the exact endpoint ID chosen does not impact the logic at all.

We must not be required to have endpoint IDs 1, 2, 3, xxx.

The exact mapping of endpoint IDs should be decoupled.

Reply:
That's what the emberAfGetClusterServerEndpointIndex function does though, doesn't it? It maps the static and dynamic endpoints into an array style index to abstract away the actual endpoint ids.

Reply:
ep is probably the wrong var name here though.

Reply:
The table size here is the count of endpoints which are suppport MicrowaveOvenControl cluster, and plus the dynamic endpoints count.  We can obtain an index from "emberAfGetClusterServerEndpointIndex" , then set/get the OpState instance by using that index.